feat(bundler): authentication support using hostRules

bilby91 opened the pull request:
Adding support for bundler credentials on docker executor using env variable approach from https://bundler.io/v1.16/bundle_config.html#CREDENTIALS-FOR-GEM-SOURCES

I've tested it with my private repositories and I was able to get bundler to run successfully and create MRs/PRs.

Closes #5023

bilby91 commented on lib/util/host-rules.ts:
```diff
@@ -18,6 +18,16 @@ export interface HostRule {
 
 let hostRules: HostRule[] = [];
 
+export function findAll({ hostType }: { hostType: string }): HostRule[] {
```
I'll wait for #5164 to get merged so I can rebase and use the tested `findAll`

bilby91 commented on lib/util/host-rules.ts:
```diff
@@ -18,6 +18,16 @@ export interface HostRule {
+export function getHost(hostRule: HostRule): string {
+  const url = URL.parse(hostRule.domainName);
+
+  return url.host || url.path;
```
@rarkins I'm not sure I got the difference between `domainName` and `hostName` on the `HostRule` type. When is one or the other used ? Should we try and support both ?

rarkins replied:
domainName is like `github.com`, hostName is like `api.github.com`. If Bundler's approach is *host* driven, then sometimes domainName might not work, but there may be no harm in trying it. `hostName` should be what you're looking for without needing any parsing.

bilby91 replied:
Do you think it would make sense to try `hostName || domainName` (parsing probably needed for `domainName` ? Also, are users expected to set the protocol when setting `hostName` ?

rarkins replied:
Both hostName and domainName are protocol-free. It's probably ok to use the logic `hostName || domainName` as nothing to lose by doing so. One example is `github.com` (domainName) vs `api.github.com` (hostName)

bilby91 replied:
Made the change so that we ue `hostName || domainName`

bilby91 commented on lib/manager/bundler/artifacts.ts:
```diff
@@ -90,6 +93,19 @@ async function getDockerTag(updateArtifact: UpdateArtifact): Promise<string> {
+        .map(term => term.toUpperCase())
+        .join('__')
+    )
+    .join('');
```
Not really, I uses join because the logic already was dealing with an array. I'll change the last `join("")` to just concat the two strings.

bilby91 commented on lib/manager/bundler/artifacts.ts:
```diff
@@ -125,11 +141,19 @@ export async function updateArtifacts(
+    const bundlerHostRulesVariables = findAll({ hostType: 'bundler' })
+      .filter(
+        hostRule =>
+          hostRule.domainName && hostRule.username && hostRule.password
```
I didn't realize that there was another key for `token`. I'll take a look at supporting token auth as well.

bilby91 commented on lib/manager/bundler/artifacts.ts:
```diff
@@ -125,11 +141,19 @@ export async function updateArtifacts(
         image: 'renovate/ruby',
         tag: await getDockerTag(updateArtifact),
         preCommands,
+        envVars: bundlerHostRulesVariables,
```
You are thinking of setting up the env variable for the child process when the child is not a docker run invocation right ? I think we should be able to support that fairly easy.

rarkins replied:
Yes, that's what I'm thinking. As they are passed via env (instead of `bundle config` commands) then it should be only a sandboxed change that doesn't "affect" the user's original env once Renovate finishes.

bilby91 replied:
Now we are passing the envVariables as `extraEnv` to `exec` and that will add the variables to the `childEnv`.

bilby91 commented on lib/util/exec/index.ts:
```diff
@@ -107,7 +107,7 @@ export async function exec(
       ...docker,
       cwd,
-      envVars: dockerEnvVars(extraEnv, childEnv),
+      envVars: dockerEnvVars(extraEnv, childEnv).concat(docker.envVars),
```
@rarkins I'm resuming my work on this one, I have my code rebased with master and most of the code review changes done.

I'm having a hard time understanding how `dockerEnvVars` should work. That function will basically extract the keys for the envVariables in `childEnv`. So, if I have some variables like this:

```
{
 "BUNDLE_GEMS__PRIVATE__COM": "some-user:some-password"
}
```

It will return an array with `["BUNDLE_GEMS__PRIVATE__COM"]`. 

Shouldn't this method generate an array like this one:

```
["BUNDLE_GEMS__PRIVATE__COM=some-user:some-password"]
```

If not the env variables are not getting passed correctly to the docker command builder. I might be missing something here or there might be another way to set them.

rarkins replied:
@zharinov can you help out on this one?

bilby91 replied:
Example test case that starts failing with the change:

```
  ● .updateArtifacts() › supports docker mode with credentials

    expect(received).toMatchSnapshot()

    Snapshot name: `.updateArtifacts() supports docker mode with credentials 1`

    - Snapshot  - 2
    + Received  + 2

    @@ -4,18 +4,18 @@
          "options": Object {
            "encoding": "utf-8",
          },
        },
        Object {
    -     "cmd": "docker run --rm --user=foobar -v \"/tmp/github/some/repo\":\"/tmp/github/some/repo\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -v \"/tmp/renovate/cache/others/go\":\"/tmp/renovate/cache/others/go\" -e GOPATH -e GOPROXY -e GONOSUMDB -e CGO_ENABLED -w \"/tmp/github/some/repo\" renovate/go bash -l -c \"git config --global url.\\\"https://some-token@github.com/\\\".insteadOf \\\"https://github.com/\\\" && go get -d ./...\"",
    +     "cmd": "docker run --rm --user=foobar -v \"/tmp/github/some/repo\":\"/tmp/github/some/repo\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -v \"/Users/bilby91/go\":\"/Users/bilby91/go\" -e GOPATH=/Users/bilby91/go -e GOPROXY=proxy.example.com -e GONOSUMDB=1 -e CGO_ENABLED=1 -w \"/tmp/github/some/repo\" renovate/go bash -l -c \"git config --global url.\\\"https://some-token@github.com/\\\".insteadOf \\\"https://github.com/\\\" && go get -d ./...\"",
```

What would be the purpose of doing `-e GOPATH` instead of `-e GOPATH=/Users/bilby91/go` ?

rarkins replied:
I think that your `GOPATH` may be getting passed through to the test case, so it ends up with a different snapshot. Need to either unset GOPATH somehow or avoid using a snapshot for that test.

bilby91 replied:
@rarkins I think I learned a new behaviour of `docker run` that I was not aware. Doing `-e MY_VAR` will pick the value from the current `env`. I thought that this was not supported and you always had to do `-e MY_VAR=my-value`. I'll get the changes done now!

bilby91 commented:
@rarkins I think it should be ready for another review. I still need to do some manual testing to verify everything is working as expected.

bilby91 commented:
@rarkins Did some manual tests using docker and non docker binary source. Both of them seem to be working fine.

rarkins commented:
Thanks, I’ll take a look soon!

bilby91 commented on lib/util/host-rules.ts:
```diff
@@ -18,6 +18,14 @@ export interface HostRule {
+}
+
+export function getAuthenticationHeaderValue(hostRule: HostRule): string {
+  return `${hostRule.username}:${hostRule.password || hostRule.token}`;
```
Yeah, the token version is wrong. Thanks for catching that (is the only piece I haven't manually tested yet). 

So, instead we should do :

```
if (hostRule.username) {
  return  `${hostRule.username}:${hostRule.password}`;
} else {
  return hostRule.token
}
```

Not sure if there is any custom logic for attaching `x-oauth-basic` in the tokens or the user is expected to add those pieces to the token variable. Can you confirm that ?

bilby91 commented on lib/util/host-rules.ts:
```diff
@@ -47,6 +55,14 @@ export interface HostRuleSearch {
+  return (
+    (!!rule.hostName || !!rule.domainName) &&
+    (!!rule.password || !!rule.token) &&
+    !!rule.username
```
My confusion here, not needed.

bilby91 commented:
@rarkins Do you want to move all new host rules functions into bundler or you think we can keep some ?

bilby91 commented:
@rarkins Do you want to move all new host rules functions into bundler or you think we can keep some ?

rarkins commented:
> @rarkins Do you want to move all new host rules functions into bundler or you think we can keep some ?

Please move all new host rules functions into the Bundler artifacts source. Each of them includes something Bundler-specific, so for now let's just keep it as Bundler logic.

bilby91 commented:
@rarkins I added support for baseURL, fixed the username/token issue and changed the method to `getDomain` instead of `getHost`. 

Also removed the new code from host-rules and move it to it's own file under `lib/manager/bundler/host-rules.ts`. Not sure if you think this is the best name for the file. I didn't add the file in the artifacts code because it would have been very difficult to test since the artifacts tests depend heavily on stubbing/mocking. 

Let me know what you think.

bilby91 commented:
@rarkins Is there anything on my end that we are waiting on ? Just double checking.

rarkins commented:
Error from CI:

```
$ tsc -p tsconfig.app.json
##[error]lib/manager/bundler/artifacts.ts(15,10): error TS2459: Module '"../../util/host-rules"' declares 'HostRule' locally, but it is not exported.
##[error]lib/manager/bundler/host-rules.ts(2,19): error TS2459: Module '"../../util/host-rules"' declares 'HostRule' locally, but it is not exported.
```

bilby91 commented:
@rarkins Sorry for the late response here. I see that you commited the fixes. Can I help with anything ?

bilby91 commented:
I'll probably have more time in the upcoming days to help!

rarkins commented:
Thanks @bilby91, I merged this now.

micheelengronne commented:
Does it also work with the `binarySource` as `global` and not `docker` ?

bilby91 commented:
@micheelengronne Should work with global. You are having issues?

micheelengronne commented:
in fact it is not with global. I don't know why but renovate wants to check my geminabox with `api/v1/gems`. I set my hostRules that way:

```
 hostRules: [
   {
     hostType: 'bundler',
     hostName: 'geminabox.domain.test',
     username: 'username',
     password: 'PASSWORD',
   }
 ]

```

And I have that in my log:

```
       "registry": "https://geminabox.domain.test",
       "err": {
         "name": "HTTPError",
         "hostname": "geminabox.domain.test",
         "method": "GET",
         "path": "/api/v1/gems/rubocop.json",
         "protocol": "https:",
         "url": "https://geminabox.domain.test/api/v1/gems/rubocop.json",
         "gotOptions": {
           "path": "/api/v1/gems/rubocop.json",
           "protocol": "https:",
           "hostname": "geminabox.domain.test",
           "hash": "",
           "search": "",
           "pathname": "/api/v1/gems/rubocop.json",
           "href": "https://geminabox.domain.test/api/v1/gems/rubocop.json",
           "headers": {
             "user-agent": "https://github.com/renovatebot/renovate",
             "hosttype": "rubygems",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate"
           },
           "hooks": {
             "beforeError": [],
             "init": [],
             "beforeRequest": [],
             "beforeRedirect": [],
             "beforeRetry": [],
             "afterResponse": []
           },
           "retry": {"methods": {}, "statusCodes": {}, "errorCodes": {}},
           "decompress": true,
           "throwHttpErrors": true,
           "followRedirect": true,
           "stream": false,
           "form": false,
           "json": true,
           "cache": false,
           "useElectronNet": false,
           "baseUrl": "https://geminabox.domain.test/",
           "method": "GET",
           "gotTimeout": {"request": 60000}
         },
         "statusCode": 401,
         "statusMessage": "Unauthorized",
         "headers": {
           "server": "nginx",
           "date": "Fri, 24 Apr 2020 13:09:29 GMT",
           "content-type": "text/html;charset=utf-8",
           "content-length": "16",
           "connection": "close",
           "x-powered-by": "geminabox 1.1.1",
           "www-authenticate": "Basic realm=\"Gem In a Box\"",
           "x-xss-protection": "1; mode=block",
           "x-content-type-options": "nosniff",
           "x-frame-options": "SAMEORIGIN",
           "set-cookie": [
             "rack.session=d0a82d19161c6b78b68645fef317c1cf57606036ee6c2e1a4f445a40484cfef5; path=/; expires=Fri, 24 Apr 2020 13:26:09 -0000; HttpOnly"
           ]
         },
         "body": "Not Authorized.\n",
         "message": "Response code 401 (Unauthorized)",
         "stack": "HTTPError: Response code 401 (Unauthorized)\n    at EventEmitter.<anonymous> (/home/renovate/.config/yarn/global/node_modules/got/source/as-promise.js:74:19)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
       },
```

micheelengronne commented:
When I remove the `hostType: bundler`, renovate detects credentials and gives a 404 as geminabox does not handle `/api/v1/gems/[:gem].json` either.

micheelengronne commented:
I removed my comments. It is off-topic. I will investigate further and open a new issue if needed.
